### PR TITLE
Import raft

### PR DESCRIPTION
**Pull request by cole-miller**
Still WIP but this is already enough to get the dqlite tests passing (on my machine, both with bundled raft and external raft). Major points:

- the new source structure is: raft sources under src/raft; the public includes raft.h, raft/uv.h, and raft/fixture.h concatenated into src/raft.h
- raft_uv is build unconditionally, as is raft_fixture
- libraft is not built separately, it all goes into libdqlite.a
- imported files are from latest raft master, 070b1b124d9c1f96553fef1c1356cde0c2e81ccf
- configure.ac and Makefile.am should be updated, we inherit some stuff from the raft versions of these files

Still todo:

- [x] get the tests to pass on GHA
- [x] update the downstream workflow and Jepsen
- [x] fix RAFT_API and visibility of raft stuff generally
- [x] import raft tests and get them passing
- [x] normalize style (e.g. spaces vs. tabs) across the codebase
- [x] test with bundled raft in CI
- [x] update dqlite.pc.in
- [ ] can we avoid building raft_fixture unless necessary?
- [ ] other things that I'm forgetting

### CODE REVIEW

**Comment by cole-miller** [comment]
I've got the clang-tidy check mostly working for patches. For some reason the identifier case check doesn't work for type names (https://discourse.llvm.org/t/clang-tidying-c-code-structcase-not-working/76863), but otherwise it seems to run as expected.

**Comment by cole-miller** [comment]
More details on what this PR does:

- The testing matrix now uses `--enable-build-raft` since that will be the first-class build configuration going forward. There is a separate CI workflow to ensure that building with external libraft still works.
- `.clang-tidy` is checked in and added to the linting workflow, but I decided not to rewrite all the problematic identifiers just yet; I'm going to do it for the whole codebase in a separate PR instead.
- The imported raft code has been autoformatted.

**Comment by cole-miller** [comment]
please test downstream

**Review by just-now — COMMENTED** [review]
Readme.md shall be updated w.r.t. recent changes.

**Comment by cole-miller** [comment]
There's definitely some flakiness in this test:

```
integration-test cluster/restart --param num_records 993 --param disk_mode 1
```

Looking into it.

File: m4/ax_pthread.m4 [hunk 1]
```diff
@@ -14,20 +14,24 @@
 #   flags that are needed. (The user can also force certain compiler
 #   flags/libs to be tested by setting these environment variables.)
 #
-#   Also sets PTHREAD_CC to any special C compiler that is needed for
-#   multi-threaded programs (defaults to the value of CC otherwise). (This
-#   is necessary on AIX to use the special cc_r compiler alias.)
+#   Also sets PTHREAD_CC and PTHREAD_CXX to any special C compiler that is
```
Comment (MathieuBordere) [inline m4/ax_pthread.m4:17]: I'm curious where you found this or what the problem was?

**Review by MathieuBordere — APPROVED** [review]
Looks good, there are 2 failing tests though, not sure if they're related, but would at least need a look.

File: m4/ax_pthread.m4 (hunk 1 above)
Comment (cole-miller) [reply m4/ax_pthread.m4:17]: I got it from [here](https://www.gnu.org/software/autoconf-archive/ax_pthread.html) (link at the bottom where it says "latest version of ax_pthread.m4"). The problem was this warning emitting when doing autoreconf -i in a fresh checkout:

```
configure.ac:12: warning: $as_echo is obsolete; use AS_ECHO(["message"]) instead
lib/m4sugar/m4sh.m4:692: _AS_IF_ELSE is expanded from...
lib/m4sugar/m4sh.m4:699: AS_IF is expanded from...
./lib/autoconf/general.m4:2249: AC_CACHE_VAL is expanded from...
./lib/autoconf/general.m4:2270: AC_CACHE_CHECK is expanded from...
m4/ax_pthread.m4:89: AX_PTHREAD is expanded from...
configure.ac:12: the top level
```

**Comment by freeekanayaka** [comment]
Hey @cole-miller now that this PR is merged, any plan to cut a new dqlite release soon? So distributions can start update their packages to the new layout. Thanks.

**Pull request merged**